openmeter/billing/service/lineservice/usagebasedline_test.go

```diff
@@ -0,0 +1,1366 @@
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.TieredPrice{
+				Mode:          plan.GraduatedTieredPrice,
+				Tiers:         testTiers,
+				MinimumAmount: lo.ToPtr(alpacadecimal.NewFromFloat(100)),
+			}),
+			lineMode: lastInPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(0),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: minimum spend",
+					Amount:                 alpacadecimal.NewFromFloat(100),
+					Quantity:               alpacadecimal.NewFromFloat(1),
+					ChildUniqueReferenceID: GraduatedMinSpendChildUniqueReferenceID,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+				},
+			},
+		})
+	})
+
+	// Maximum spend
+	t.Run("tiered graduated, first price, usage eq max spend", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.TieredPrice{
+				Mode:          plan.GraduatedTieredPrice,
+				Tiers:         testTiers,
+				MaximumAmount: lo.ToPtr(alpacadecimal.NewFromFloat(100)),
+			}),
+			lineMode: lastInPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(5),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: flat price for tier 1",
+					Amount:                 alpacadecimal.NewFromFloat(100),
+					Quantity:               alpacadecimal.NewFromFloat(1),
+					ChildUniqueReferenceID: GraduatedFlatPriceChildUniqueReferenceID,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+				},
+			},
+		})
+	})
+
+	t.Run("tiered graduated, first price, usage above max spend, max spend is not at tier boundary ", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.TieredPrice{
+				Mode:          plan.GraduatedTieredPrice,
+				Tiers:         testTiers,
+				MaximumAmount: lo.ToPtr(alpacadecimal.NewFromFloat(125)),
+			}),
+			lineMode: lastInPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(7),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: flat price for tier 2",
+					Amount:                 alpacadecimal.NewFromFloat(150),
+					Quantity:               alpacadecimal.NewFromFloat(1),
+					ChildUniqueReferenceID: GraduatedFlatPriceChildUniqueReferenceID,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+					Discounts: []billingentity.LineDiscount{
+						{
+							Description: lo.ToPtr("Maximum spend discount for charges over 125"),
+							Amount:      alpacadecimal.NewFromFloat(25),
+							Type:        lo.ToPtr(billingentity.MaximumSpendLineDiscountType),
+							Source:      billingentity.CalculatedLineDiscountSource,
+						},
+					},
+				},
+			},
+		})
+	})
+}
+
+func TestTieredVolumeCalculation(t *testing.T) {
+	testTiers := []plan.PriceTier{
```

Comment:
On volume based each unit is billed on the highest achieved discount, see:
https://openmeter.io/blog/tiered-pricing-model#example-comparison

Reply:
To match the plan's notion we are using the following terminology:
Tiered/Volume based prices are what the blog names Volume-based pricing
Tiered/Graduating based prices are what the blog names Tiered-based pricing